AD incompressible navier stokes

lindsayad opened the pull request:
Last piece needed in order to move Baldr to using upstream MOOSE

lindsayad committed:
Add AD incompressible navier stokes

lindsayad committed:
Generalize ADKernelValue and ADKernelGrad

lindsayad committed:
Create INSADTauMaterial

lindsayad committed:
Add ADKernelSUPG class

lindsayad committed:
Error if trying to use seconds with displacements

lindsayad committed:
Reimplement INSADMass

lindsayad committed:
Create INSADMaterial

lindsayad committed:
Reimplement INSADMomentum*

lindsayad committed:
Reimplement INSADMomentum

lindsayad committed:
Move INSADMomentumAdvection to INSADMomentum

lindsayad committed:
Add adCoupledVectorDot

lindsayad committed:
Shuffling around using declarations

lindsayad committed:
Cleaning up INSAD*Material

lindsayad committed:
Remove INSAD*Base classes

lindsayad committed:
Everything compiles

lindsayad committed:
Same results with ad_lid_driven but slow

lindsayad committed:
Make ADTimeKernel use ADKernelValue and ADDiffusion use ADKernelGrad

lindsayad committed:
Update HeatConduction module for changes in ADDiffusion and ADTimeKernel

lindsayad committed:
Remove INSADTemperatureTimeDerivative in favor of existing ADHeatConductionTimeDerivative

lindsayad committed:
Update NavierStokes module for changes in ADTimeKernel

lindsayad committed:
Update lid_driven input files. Tests pass

lindsayad committed:
Add VectorVariableComponent aux kernel

lindsayad committed:
Debugging difference between hand and ad. Difference is in SUPG

lindsayad committed:
Include grad_p in the momentum strong form

lindsayad committed:
Add alpha and make sure to index _velocity by _qp

lindsayad committed:
Reverse order of _grad_test and _velocity because of libmesh #2062

lindsayad committed:
Finish adding compatible exodiff ad tests

lindsayad committed:
Add Jacobian tests

lindsayad committed:
Overriding precomputeQpResidual in ADMatDiffusion

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/301796/) on 3590adb wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13018/site)

This comment will be updated on new commits.

lindsayad committed:
Dont add displacements as coupled vars to Material

lindsayad committed:
Include Assembly in all the ADKernel* classes

lindsayad committed:
Loosen jacobian tolerance slightly in ad_heat_source_bar

ADDiffusion and ADMatDiffusion now inherit from ADKernelGrad and
the floating point operations have been changed up slightly...
enough to aggravate this test

lindsayad committed:
ADDiffusionPrecompute object no longer needed

lindsayad committed:
Clean up ADResidual types

lindsayad committed:
Document VectorVariableComponentAux

lindsayad committed:
SQA test documentation

lindsayad committed:
Split out INSAD classes

lindsayad committed:
Get sign correct for gravity term

lindsayad committed:
Add NavierStokes documentation

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/302513/) on f8d3a8a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13018/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13018/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 19a3d8261784a922ecb04f45154dc36d715f4138`

lindsayad committed:
Remove *ConvectedMesh* for now. Need to be converted to vector

aeslaughter committed:
Add heat conduction to documentation

dschwen commented on framework/src/kernels/ADKernelSUPG.C:
```diff
@@ -0,0 +1,182 @@
+  const unsigned int n_test = _grad_test.size();
+  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+  {
+    const auto value = precomputeQpStrongResidual() * _ad_JxW[_qp] * _ad_coord[_qp] * _tau[_qp];
```
shouldn't the tau be on the other term?

lindsayad replied:
You mean inside the `_i` loop? It's only a function of `_qp`

dschwen replied:
Oh, I see. I thought SUPG is

(test + tau * grad_test * v * adv) * strong_residual

But I'm not an SUPG expert by any stretch of the definition of the word "expert" :-D

lindsayad replied:
> (test + tau * grad_test * v * adv) * strong_residual

This is almost correct. For a general transient advection-diffusion problem, 

u_dot + v * grad_u + grad * (-kappa * grad_u) - f = 0

the overall weak form would look like:

(test, u_dot) + (test, v * grad_u) + (grad_test, kappa * grad_u) - (test, f) - <test, n * kappa * grad_u> + (tau * v * grad_test, u_dot + v * grad_u + grad * (-kappa * grad_u) - f) = 0

So the difference is in the treatment of second order terms...those terms are integrated by parts when applying the standard test function, but not when applying the Petrov-Galerkin modificiation, tau * v * grad_test.

So in this design, an `ADKernelSUPG` derived object is only meant to be used to add the contribution from the Petrov-Galerkin modification. Other kernels should supply the standard test function contributions.

jwpeterson commented on framework/include/utils/MathUtils.h:
```diff
@@ -77,6 +77,35 @@ addScaled(const T & scalar, const NumericVector<T2> & numeric_vector, NumericVec
+          typename T2,
+          template <typename> class W,
+          template <typename> class W2,
+          typename std::enable_if<std::is_same<typename W<T>::index_type,
```
Yo dawg, heard you like nested template parameters!

lindsayad replied:
🤣

jwpeterson reviewed (dismissed):
I won't pretend to understand all the macros and template magic here, but this is definitely an awesome capability!

lindsayad committed:
Add heat_conduction dependency to navier_stokes in modules.mk. Closes #13025
